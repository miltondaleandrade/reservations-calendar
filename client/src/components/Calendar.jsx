/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import styles from './styles.css';

class Calendar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      year: this.props.selectedDate.getFullYear(),
      month: this.props.selectedDate.getMonth(),
    };

    this.monthNames = ['January', 'February', 'March', 'April', 'May', 'June', 'July',
      'August', 'September', 'October', 'November', 'December'];
    this.dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    this.closedDaysIndex = [];
    for (let i = 0; i < this.dayNames.length; i += 1) {
      if (this.props.openHours[this.dayNames[i]].length === 0) {
        this.closedDaysIndex.push(i);
      }
    }
  }

  generateCalendar(year, month) {
    const firstDayOfMonth = new Date(year, month).getDay();
    const numOfDays = 32 - new Date(year, month, 32).getDate();

    const numOfDaysPrev = 32 - new Date(year, month - 1, 32).getDate();
    const lastDayofPrevMonth = new Date(year, month - 1, numOfDaysPrev).getDay();

    const currentDate = new Date();
    const currentDay = currentDate.getDate();
    const currentMonth = Number(currentDate.getMonth());
    const currentYear = currentDate.getFullYear();

    let date = 1;
    const calendarRows = [];

    for (let r = 0; r < 5; r += 1) {
      const daysInRow = [];

      for (let i = 0; i < 7; i += 1) {
        if (r === 0 && i < firstDayOfMonth) {
          const prevDate = numOfDaysPrev - (lastDayofPrevMonth - i);
          const greyCell = <td className={styles.greyCell} key={`prev${prevDate}`}>{prevDate}</td>;
          daysInRow.push(greyCell);
        } else if (date > numOfDays) {
          const nextDate = date - numOfDays;
          const futureGreyCell = (
            <td 
              className={styles.futureGreyCell}
              key={`next${nextDate}`}
              month={month}
              year={year}
              day={nextDate}
              onClick={this.props.handleDayClick}
            >
              {nextDate}
            </td>
          );
          daysInRow.push(futureGreyCell);
          date += 1;
        } else {
          if (this.closedDaysIndex.includes(i)) {
            const greyCell = <td className={styles.greyCell} key={`current${date}`}>{date}</td>;
            daysInRow.push(greyCell);
          } else if (month === currentMonth && date < currentDay) {
            const greyCell = <td className={styles.greyCell} key={`current${date}`}>{date}</td>;
            daysInRow.push(greyCell);
          } else {
            const day = date < 10 ? `0${date}` : date;
            let calendarMonth = Number(month) + 1;
            calendarMonth = calendarMonth < 10 ? `0${calendarMonth}` : calendarMonth;
            const dateString = (new Date(`${year}-${calendarMonth}-${day}T10:20:30Z`)).toDateString();
            const filledCell = (
              <td
                className={dateString === this.props.selectedDate.toDateString() ? styles.selectedCell : styles.filledCell}
                key={`current${date}`}
                month={month}
                year={year}
                day={date}
                onClick={this.props.handleDayClick}
              >
                {date}
              </td>
            );
            daysInRow.push(filledCell);
          }
          date += 1;
        }
      }
      calendarRows.push(daysInRow);
    }
    return calendarRows;
  }

  handleLeftArrowClick(event) {
    if (this.state.month !== 0) {
      this.setState({
        month: this.state.month - 1,
      });
    } else {
      this.setState({
        year: this.state.year - 1,
        month: 11,
      });
    }
  }

  handleRightArrowClick(event) {
    if (this.state.month !== 11) {
      this.setState({
        month: this.state.month + 1,
      });
    } else {
      this.setState({
        year: this.state.year + 1,
        month: 0,
      });
    }
  }

  render() {
    const calendarPage = this.generateCalendar(this.state.year, this.state.month);
    return (
      <div className={styles.calendar}>
        <div className={styles.captionRow}>
          <div className={styles.arrowButton}>
            <div className={styles.leftArrow} onClick={this.handleLeftArrowClick.bind(this)}><i className="fas fa-angle-left" /></div>
          </div>
          <div className={styles.monthYearContainer}>
            <div className={styles.monthYear}>
              {this.monthNames[this.state.month]}
              {' '}
              {this.state.year}
            </div>
          </div>
          <div className={styles.arrowButton}>
            <div className={styles.rightArrow} onClick={this.handleRightArrowClick.bind(this)}><i className="fas fa-angle-right" /></div>
          </div>
        </div>
        <table>
          <thead className={styles.tableHead}>
            <tr>
              <td className={styles.weekday}>Su</td>
              <td className={styles.weekday}>Mo</td>
              <td className={styles.weekday}>Tu</td>
              <td className={styles.weekday}>We</td>
              <td className={styles.weekday}>Th</td>
              <td className={styles.weekday}>Fr</td>
              <td className={styles.weekday}>Sa</td>
            </tr>
          </thead>
          <tbody className={styles.tbody} id="tbody">
            <tr>
              {calendarPage[0].map((day) => day)}
            </tr>
            <tr>
              {calendarPage[1].map((day) => day)}
            </tr>
            <tr>
              {calendarPage[2].map((day) => day)}
            </tr>
            <tr>
              {calendarPage[3].map((day) => day)}
            </tr>
            <tr>
              {calendarPage[4].map((day) => day)}
            </tr>
          </tbody>
        </table>
      </div>
    );
  }
}

export default Calendar;
