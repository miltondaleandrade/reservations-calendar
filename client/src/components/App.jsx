import React from 'react';
import axios from 'axios';
import Calendar from './Calendar.jsx';

import styles from './styles.css';

class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      date: this.dateString,
      dateUTC: new Date(),
      time: undefined,
      partySize: undefined,
      displayCalendar: false,
      dayOfWeek: null,
      restaurantData: null,
    };

    this.date = new Date();
    this.dateString = this.date.toDateString();
    this.dayNames = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    this.partySizes = ['1 person'];

    this.handleChange = this.handleChange.bind(this);
    this.handleDayClick = this.handleDayClick.bind(this);
    this.displayCalendar = this.displayCalendar.bind(this);
  }

  componentDidMount() {
    this.setState({
      dateUTC: new Date(),
      date: this.date.toDateString(),
      dayOfWeek: this.date.getDay(),
    });
    this.getData();
  }

  static onMouseDown(event) {
    // eslint-disable-next-line no-param-reassign
    event.target.style.background = '#be2020';
  }

  getData() {
    axios.get('/reservations/')
      .then((response) => {
        const dayOfWeek = this.date.getDay();
        const dayTimes = response.data.openHours[this.dayNames[dayOfWeek]];
        this.setState({
          restaurantData: response.data,
          time: dayTimes.includes('7:00 pm') ? '7:00 pm' : dayTimes[5],
        });
        return response;
      })
      .then(() => {
        this.generatePartySizes(this.partySizes);
      })
      .catch((err) => {
        console.log('Error contating server! ', err);
      });
  }

  generatePartySizes(array) {
    const { restaurantData } = this.state;
    const maxSize = restaurantData.maxPartySize;
    for (let i = 2; i <= maxSize; i += 1) {
      const size = `${i} people`;
      array.push(size);
    }
    this.setState({
      partySize: '2 people',
    });
  }

  displayCalendar() {
    const {
      displayCalendar,
    } = this.state;
    this.setState({
      displayCalendar: !displayCalendar,
    });
  }

  handleDayClick(event) {
    let month = Number(event.target.getAttribute('month')) + 1;
    const year = event.target.getAttribute('year');
    let day = event.target.getAttribute('day');
    if (day < 10) {
      day = `0${day}`;
    }
    if (month < 10) {
      month = `0${month}`;
    }
    const dateUTC = new Date(`${year}-${month}-${day}T10:20:30Z`);
    const dateString = dateUTC.toDateString();
    this.setState({
      date: dateString,
      displayCalendar: false,
      dateUTC,
      dayOfWeek: dateUTC.getDay(),
    });
  }

  handleChange(event) {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  render() {
    const {
      date, time, partySize, displayCalendar, dateUTC, dayOfWeek, restaurantData,
    } = this.state;
    return (
      <div className={styles.reservationMod} id="reservationBox">
        <div>
          <h2 id={styles.header}>Make a Reservation</h2>
        </div>
        <div>
          <input
            id={styles.calendarSelect}
            defaultValue={this.dateString}
            value={date}
            onClick={this.displayCalendar}
            readOnly
          />
        </div>
        <div id={styles.selectRow}>
          <select id={styles.time} name="time" value={time} onChange={this.handleChange}>
            {restaurantData && restaurantData.openHours[this.dayNames[dayOfWeek]].map((resTime) => (
              <option
                value={resTime}
                key={resTime}
              >
                {resTime}
              </option>
            ))}
          </select>
          <select id={styles.party} name="partySize" value={partySize} onChange={this.handleChange}>
            {partySize && this.partySizes.map((party) => (
              <option
                value={party}
                key={party}
              >
                {party}
              </option>
            ))}
          </select>
        </div>
        <div>
          <button id={styles.button} type="button" onMouseDown={App.onMouseDown}>Find a Table</button>
        </div>
        <div>
          {displayCalendar ? (
            <Calendar
              selectedDate={dateUTC}
              openHours={restaurantData.openHours}
              dayNames={this.dayNames}
              handleDayClick={this.handleDayClick}
            />
          ) : null}
        </div>
      </div>

    );
  }
}

export default App;
