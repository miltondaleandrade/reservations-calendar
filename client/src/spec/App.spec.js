import React from 'react';
import { shallow, mount } from 'enzyme';
import App from '../components/App';
import Calendar from '../components/Calendar';

describe('App', () => {

  it('should mount the App component', () => {
    const wrap = shallow(<App />);
    expect(wrap).toMatchSnapshot();
  });

  it('should contain an <h2> tag with recognizable text', () => {
    const wrap = shallow(<App />);
    const h2 = wrap.find('h2');
    const result = h2.text();
    expect(result).toBe('Make a Reservation');
  });

  it('should include a button that says "Find a Table"', () => {
    const wrap = shallow(<App />);
    const button = wrap.find('button');
    const result = button.text();
    expect(result).toEqual('Find a Table');
  });

  it('should load a date selector showing the current date', () => {
    const wrap = shallow(<App />);
    const input = wrap.find('input');
    const result = input.text();
    const currentDate = new Date();
    const currentDateString = currentDate.toDateString();
    expect(result).toEqual(currentDateString);
  });

  it('should display a party size selector that contains all party options"', () => {
    const wrap = shallow(<App />);
    const partySelector = wrap.find('select').at(1);
    const result = partySelector.text();
    expect(result).toEqual('1 person2 people3 people4 people5 people6 people');
  });

  it('should not display the calendar when App is first mounted', () => {
    const wrap = shallow(<App />);
    expect(wrap.state('displayCalendar')).toEqual(false);
  });

  it('should change the displayCalendar state of App when the date input is clicked', () => {
    const wrap = shallow(<App />);
    wrap.find('input').simulate('click');
    expect(wrap.state('displayCalendar')).toEqual(true);
  });

  it('should display the calendar when the displayCalendar state of App is updated', () => {
    const wrap = shallow(<App />);
    wrap.setState({ displayCalendar: true });
    expect(wrap.containsMatchingElement(<Calendar />)).toBeTruthy;
  });

  it('should update the selected date when a calendar date is clicked', () => {
    const wrap = mount(<App />);
    wrap.setState({ displayCalendar: true });
    const dateCell = wrap.find('td').at(9);
    dateCell.simulate('click');
    expect(wrap.state('date')).toEqual('Sat Feb 01 2020');
  });

  it('should stop showing the calender once a calendar date is clicked', () => {
    const wrap = mount(<App />);
    wrap.setState({ displayCalendar: true });
    const filledCell = wrap.find('td').at(9);
    filledCell.simulate('click');
    expect(wrap.state('displayCalendar')).toEqual(false);
  });

  it('should scroll between months when calendar arrows are clicked', () => {
    const wrap = shallow(<Calendar />);
    const rightArrow = wrap.find('div').at(8);
    wrap.setState({ month: 0 });
    rightArrow.simulate('click');
    expect(wrap.state('month')).toEqual(1);
  });
});



//it should update the selected date when a calendar date is clicked

//the calendar should stop showing after a date is clicked

//figure out how to grab by elementID when using CSS mods - how to use find when multiple of the same type

//figure out how to check the default value of an input in a test
