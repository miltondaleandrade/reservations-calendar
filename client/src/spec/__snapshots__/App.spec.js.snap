// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should mount the App component 1`] = `ShallowWrapper {}`;
